compiler/sem/lookups.nim

```diff
@@ -674,24 +676,22 @@ proc errorUndeclaredIdentifierWithHint*(
   ## imports
   result = errorSym(c, n, c.config.newError(
     n,
-    SemReport(
-      kind: rsemUndeclaredIdentifier,
-      potentiallyRecursive: c.recursiveDep.len > 0,
+    PAstDiag(
+      kind: adSemUndeclaredIdentifier,
+      recursiveDeps: c.recursiveDep,
       spellingCandidates: candidates,
-      str: name)))
+      name: name)))
 
-  if c.recursiveDep.len > 0:
-    c.recursiveDep.setLen 0
+  c.clearRecursiveDeps()
 
 proc errorAmbiguousUseQualifier(
     c: PContext; ident: PIdent, n: PNode, candidates: seq[PSym]
   ): PSym =
   ## create an error symbol for an ambiguous unqualified lookup
-  var rep = reportSym(rsemAmbiguousIdent, candidates[0])
-  for i, candidate in candidates.pairs:
-    rep.symbols.add candidate
-
-  let err = c.config.newError(n, rep)
+  let
+    diag = PAstDiag(kind: adSemAmbiguousIdentWithCandidates,
+                    candidateSyms: candidates)
```

Comment:
Since `candidates` is consumed here, you can use `sink` for the parameter. That gets around an unnecessary copy.

Reply:
done